File: programmers/해시/전화번호 목록.py
def solution(phone_book):
    answer = True

    phone_book.sort()
    # r_phone_book = reversed(phone_book)

    zip_phone = list(zip(phone_book, phone_book[1:]))

    # print(zip_phone)

    for p1, p2 in zip_phone:
        if p2.startswith(p1):
            return False

    # 효율성 테스트 실패
    # for i in range(len(phone_book)):
    #     for j in range(i+1,len(phone_book)):
    #         if phone_book[j].startswith(phone_book[i]) == True:
    #             answer = False
    #             break

    return answer

File: programmers/해시/전화번호목록.java
import java.util.*;
class Solution {
    public boolean solution(String[] phone_book) {

        // boolean answer = true;

        // Arrays.sort(phone_book); 불필요!!
        // System.out.println(Arrays.toString(phone_book));
        Map<String, String> map = new HashMap<String, String>();

        for(String tmp:phone_book) {
        	map.put(tmp, tmp); // key = vaule
        }

        for(String str:phone_book) {
        	for(int i=0; i<str.length(); i++) {
        		String tmp = str.substring(0, i);
                System.out.println(tmp);
        		if(map.containsKey(tmp)) return false;
        	}
        }
        return true;

     // 시간초과....
        // for(int i=0; i<phone_book.length; i++) {
        //     for (int j=i+1; j<phone_book.length; j++){
        //         if(phone_book[i].startsWith(phone_book[j]))
        //             return false;
        //         if(phone_book[j].startsWith(phone_book[i]))
        //             return false;
        //     }
        // }
        // return true;


        // 정답
        // Arrays.sort(phone_book);
        // boolean result = true;
        // for (int i=0; i<phone_book.length-1; i++) {
        //     if (phone_book[i+1].startsWith(phone_book[i])) {
        //         result = false;
        //         break;
        //     }
        // }
        // return result;
    }
}
